<launch>
  <!-- ***************************************************************************************************** -->
  <!-- stop_factor_search -->
  <arg name="ns" default="/"/>
  <arg name="execution_period" default="0.04"/>
  <arg name="detection_area_width" default="3.0"/>
  <arg name="extend_area_size" default="7.0"/>
  <arg name="search_distance" default="50.0"/>
  <arg name="points_topic" default="/points_no_ground"/>
  <arg name="sub_wps_topic" default="/base_waypoints"/>
  <arg name="points_threshold" default="1"/>
  <arg name="stopline_stop_duration" default="6.0"/>
  <arg name="map_frame" default="map"/>

  <group ns="$(arg ns)">
    <node pkg="stop_factor_search" type="stop_factor_search" name="stop_factor_search" output="log">
      <param name="execution_period" value="$(arg execution_period)"/>
      <param name="detection_area_width" value="$(arg detection_area_width)"/>
      <param name="extend_area_size" value="$(arg extend_area_size)"/>
      <param name="search_distance" value="$(arg search_distance)"/>
      <param name="points_topic" value="$(arg points_topic)"/>
      <param name="sub_wps_topic" value="$(arg sub_wps_topic)"/>
      <param name="map_frame" value="$(arg map_frame)"/>
      <param name="points_threshold" value="$(arg points_threshold)"/>
      <param name="stopline_stop_duration" value="$(arg stopline_stop_duration)"/>
    </node>

    <!-- rosrun rosservice rosservice call /[node name]/set_logger_level ros.[package name] DEBUG -->
    <!--<node pkg="rosservice" type="rosservice" name="stop_factor_search_set_logger_level_debug" args="call stop_factor_search/set_logger_level ros.stop_factor_search DEBUG"/>
    -->
  </group>

  <!-- ***************************************************************************************************** -->
  <!-- ovp_velocity_planner -->
  <arg name="output" default="log"/>

  <!-- debug flags -->
  <arg name="show_debug_info" default="False"/>
  <arg name="show_debug_info_all" default="False"/>
  <arg name="show_figure" default="False"/>

  <!-- interface topic name -->
  <arg name="sub_waypoints_name" default="stop_factor_search/waypoints"/>
  <arg name="pub_waypoints_name" default="/final_waypoints"/>

  <!-- planning parameters -->
  <arg name="max_velocity" default="5.53"/>
  <arg name="max_accel" default="0.3"/>
  <arg name="min_decel" default="-1.0"/>
  <arg name="max_acc_jerk" default="0.1"/>
  <arg name="min_dec_jerk_nominal" default="-0.1"/>
  <arg name="min_dec_jerk_urgent" default="-1.5"/>
  <arg name="replan_vel_deviation" default="5.53"/>
  <arg name="engage_velocity" default="0.35"/>
  <arg name="jerk_planning_span" default="0.1"/>
  <arg name="stop_dist_mergin" default="0.55"/>
  <!-- use this velocity as command to engage vehicle speed when stopping -->
  <arg name="engage_acceleration" default="0.0"/>
  <!-- use this acceleration as command to engage vehicle speed when stopping -->
  <arg name="stopping_speed" default="0.0"/>
  <!-- planning assuming this speed is stop speed -->
  <arg name="extract_ahead_dist" default="50.0"/>
  <arg name="extract_behind_dist" default="3.0"/>
  <arg name="resample_num" default="10"/>

  <arg name="stop_dist_not_to_drive_vehicle" default="1.5"/>
  <arg name="enable_to_publish_emergency" default="true"/>

  <arg name="emergency_flag_vel_thr_kmph" default="3.0"/>
  <arg name="large_jerk_report" default="-1.4"/>

  <group ns="$(arg ns)">
    <node pkg="ovp_velocity_planner" type="velocity_planner_node" name="velocity_planner" output="$(arg output)">
      <param name="show_debug_info" value="$(arg show_debug_info)"/>
      <param name="show_debug_info_all" value="$(arg show_debug_info_all)"/>
      <param name="show_figure" value="$(arg show_figure)"/>

      <param name="sub_waypoints_name" value="$(arg sub_waypoints_name)"/>
      <param name="pub_waypoints_name" value="$(arg pub_waypoints_name)"/>

      <param name="max_velocity" value="$(arg max_velocity)"/>
      <param name="max_accel" value="$(arg max_accel)"/>
      <param name="min_decel" value="$(arg min_decel)"/>
      <param name="max_acc_jerk" value="$(arg max_acc_jerk)"/>
      <param name="min_dec_jerk_nominal" value="$(arg min_dec_jerk_nominal)"/>
      <param name="min_dec_jerk_urgent" value="$(arg min_dec_jerk_urgent)"/>
      <param name="replan_vel_deviation" value="$(arg replan_vel_deviation)"/>
      <param name="engage_velocity" value="$(arg engage_velocity)"/>
      <param name="stopping_speed" value="$(arg stopping_speed)"/>
      <param name="extract_ahead_dist" value="$(arg extract_ahead_dist)"/>
      <param name="extract_behind_dist" value="$(arg extract_behind_dist)"/>
      <param name="resample_num" value="$(arg resample_num)"/>
      <param name="jerk_planning_span" value="$(arg jerk_planning_span)"/>
      <param name="stop_dist_mergin" value="$(arg stop_dist_mergin)"/>
      <param name="emergency_flag_vel_thr_kmph" value="$(arg emergency_flag_vel_thr_kmph)"/>
      <param name="stop_dist_not_to_drive_vehicle" value="$(arg stop_dist_not_to_drive_vehicle)"/>
      <param name="enable_to_publish_emergency" value="$(arg enable_to_publish_emergency)"/>
      <param name="large_jerk_report" value="$(arg large_jerk_report)"/>
    </node>
  </group>

  <!-- ***************************************************************************************************** -->
  <!-- pure_pursuit_rev2 -->
  <!-- <arg name="ns" default="trajectory_follower"/> -->
  <arg name="uses_lerp" default="True"/>
  <arg name="publish_twist_cmd" default="False"/>
  <arg name="publish_ctrl_cmd" default="True"/>
  <arg name="reverse_minimum_lookahead_distance" default="7.0"/>
  <arg name="control_period" default="0.033"/>
  <arg name="velocity_delay_compensation_time" default="0.17"/>

  <group ns="$(arg ns)">
    <!-- rosrun waypoint_follower pure_pursuit -->
    <node pkg="rostopic" type="rostopic" name="config_waypoint_follower_rostopic" args="pub -l /config/waypoint_follower autoware_config_msgs/ConfigWaypointFollower
        '{ header: auto,
           param_flag: 0,
           velocity: 1,
           lookahead_distance: 4,
           lookahead_ratio: 2.2,
           minimum_lookahead_distance: 2.5,
           displacement_threshold: 0.0,
           relative_angle_threshold: 0.0
         }' " />
    <node pkg="pure_pursuit" type="pure_pursuit_rev2" name="pure_pursuit" output="log">
      <param name="uses_lerp" value="$(arg uses_lerp)"/>
      <param name="publish_twist_cmd" value="$(arg publish_twist_cmd)"/>
      <param name="publish_ctrl_cmd" value="$(arg publish_ctrl_cmd)"/>
      <param name="reverse_minimum_lookahead_distance" value="$(arg reverse_minimum_lookahead_distance)"/>
      <param name="control_period" value="$(arg control_period)"/>
      <param name="velocity_delay_compensation_time" value="$(arg velocity_delay_compensation_time)"/>
      <remap from="/ctrl_cmd" to="/pure_pursuit/control_cmd" />
    </node>

    <!-- rosrun rosservice rosservice call /pure_pursuit/set_logger_level ros.pure_pursuit DEBUG -->
    <!--<node pkg="rosservice" type="rosservice" name="pure_pursuit_set_logger_level_debug" args="call pure_pursuit/set_logger_level ros.pure_pursuit DEBUG"/>
    -->
  </group>

  <!-- ***************************************************************************************************** -->
  <!-- twist_gate -->
  <arg name="use_decision_maker" default="false" />

  <node pkg="twist_gate" type="twist_gate" name="twist_gate" output="log">
    <param name="use_decision_maker" value="$(arg use_decision_maker)" />
    <remap from="/lon_ctrl_cmd" to="/velocity_controller/control_cmd" />
    <remap from="/lat_ctrl_cmd" to="/pure_pursuit/control_cmd" />
  </node>

  <!-- ***************************************************************************************************** -->
</launch>
